### ISSUE TITLE
Clean up Elixir 1.3 compile warnings

### ISSUE THREAD

**Issue by barruumrex** [rrrene/credo#93]
Elixir v1.3.0-rc.0 has a new compiler feature that warns when variables are bound within a conditional statement. While Credo doesn't appear to have any actual unsafe variable bindings, it does regularly rebind variables inside conditionals. This causes the compiler to vomit uncontrollably. 

In order to be polite to end users, we should clean up these warnings.

**Comment by barruumrex** [comment]
First pass attempt at cleaning up the warnings is here https://github.com/rrrene/credo/pull/94. 

I need to make a second pass and add in some tests to make sure some of the larger refactorings didn't accidentally change anything.

**Comment by barruumrex** [comment]
Rebased off of release-0.4.0 https://github.com/rrrene/credo/pull/96

**Comment by rrrene** [comment]
Great. I reverted/refactored small parts of your PR. The CLI refactoring did indeed miss the cases where a CLI switch should supply a value rather than `true/false`. However, I liked the idea to transform the CLI config in a "Plug like" fashion. I decided to convert the Keyword list we get from `OptionParser` to a `Map` so we can more easily pattern match against the provided switches.

Overall, I'd say a very successful PR, which improved the code quality. Thanks! :tada:
